### PR TITLE
Appengine-deploy action

### PR DESCRIPTION
**Pull request by averikitsch**

### CODE REVIEW

File: .github/workflows/appengine-deploy.yml [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+    - uses: actions/setup-node@master
+      with:
+        node-version: 12.x
+
+    - name: npm install
+      run: npm install
+      working-directory: ./appengine-deploy
+
+    - name: npm lint
```
Comment (craigdbarber) [inline .github/workflows/appengine-deploy.yml:21]: Should also include a npm test invocation.

File: .github/workflows/setupGcloudSDK.yml [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+      working-directory: ./setupGcloudSDK
+      run: npm install
+
+    - name: Lint
+      working-directory: ./setupGcloudSDK
+      run: npm run lint
+
+    - name: Unit tests
+      working-directory: ./setupGcloudSDK
+      run: npm test
```
Comment (craigdbarber) [inline .github/workflows/setupGcloudSDK.yml:28]: Missing trailing '\n'

File: setupGcloudSDK/src/index.ts [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+  if (!extPath) {
+    throw new Error(`Failed to download release, url: ${url}`);
+  }
+
+  // Install the downloaded release into the github action env
+  return await installUtil.installGcloudSDK(version, extPath);
+}
+
+/**
+ * Parses the service account string into JSON
```
Comment (craigdbarber) [inline setupGcloudSDK/src/index.ts:129]: Missing trailing '.'

File: setupGcloudSDK/src/index.ts [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+ *
+ * @param serviceAccountKey The service account key used for authentication.
+ * @returns ServiceAccountKey object
```
Comment (craigdbarber) [inline setupGcloudSDK/src/index.ts:132]: Missing description for return value.

File: setupGcloudSDK/src/index.ts (hunk 3 above)
Comment (craigdbarber) [reply setupGcloudSDK/src/index.ts:129]: Here and in a couple other comment headers in this file.

File: setupGcloudSDK/src/index.ts [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+}
+
+/**
+ * Authenticates the gcloud tool using a service account key
+ *
+ * @param serviceAccountKey The service account key used for authentication.
+ * @returns exit code
+ */
+export async function authenticateGcloudSDK(
+  serviceAccountKey: string,
```
Comment (craigdbarber) [inline setupGcloudSDK/src/index.ts:152]: Remove trailing ','

**Comment by craigdbarber** [comment]
Very cool! I'm excited to see these workflows all light up green once we've added your CI infra :)

File: .github/workflows/appengine-deploy.yml (hunk 1 above)
Comment (averikitsch) [reply .github/workflows/appengine-deploy.yml:21]: I don't have any unit tests, only the integration tests. The setupGcloudSDK has all the unit tests. Do you have guidance on adding unit tests?

File: .github/workflows/setupGcloudSDK.yml (hunk 2 above)
Comment (averikitsch) [reply .github/workflows/setupGcloudSDK.yml]: done

File: setupGcloudSDK/src/index.ts (hunk 3 above)
Comment (averikitsch) [reply setupGcloudSDK/src/index.ts]: done

File: setupGcloudSDK/src/index.ts (hunk 4 above)
Comment (averikitsch) [reply setupGcloudSDK/src/index.ts]: done

File: setupGcloudSDK/src/index.ts (hunk 5 above)
Comment (averikitsch) [reply setupGcloudSDK/src/index.ts:151]: Our linter wants this trailing ','

**Comment by Bernix01** [comment]
any update on this?

File: .github/workflows/appengine-deploy.yml (hunk 1 above)
Comment (craigdbarber) [reply .github/workflows/appengine-deploy.yml:21]: Gotcha. As long as the logical units have test coverage, housing them in setupGcloudSDK makes sense.

File: appengine-deploy/README.md [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+-->
+# appengine-deploy
+
+This action deploys your source code to App Engine and makes the URL available to
+later build steps via outputs. This allows you to parameterize your App Engine
+deployments.
+
+**Note** This action will install gcloud in the background if not using in with
```
Comment (craigdbarber) [inline appengine-deploy/README.md:22]: s/**Note**/**Note:**

File: appengine-deploy/README.md [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+the [`setup-gcloud` action](../setup-gcloud/README.md).
+
+## Prerequisites
+
+This action requires Google Cloud credentials that are authorized to deploy an
+App Engine Application. See the Authorization section below for more information.
```
Comment (craigdbarber) [inline appengine-deploy/README.md:28]: Consider adding a link to App Engine documentation here.

File: appengine-deploy/README.md [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+- `url`: The URL of your App Engine Application.
+
+## Authorization
+
+There are a few ways to authenticate this action. The caller must have
+permissions to access the secrets being requested.
+
+Roles needed:
+
+- Storage Admin (`roles/compute.storageAdmin`):to upload files
```
Comment (craigdbarber) [inline appengine-deploy/README.md:81]: Missing a space after the ':'

File: appengine-deploy/README.md [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+- App Engine roles:
+  - Deployer (`roles/appengine.deployer`): Can deploy but can not promote
+  - Admin (`roles/appengine.appAdmin`): Can manage all App Engine resources (not recommended)
+
+*Note:* An owner will be needed to create the App Engine application
+
+### Used with setup-gcloud
+
+You can provide credentials using the [setup-gcloud][setup-gcloud] action,
+however you must provide your Project ID to the `appeninge-deploy` action:
```
Comment (craigdbarber) [inline appengine-deploy/README.md:92]: s/appeninge/appengine

File: appengine-deploy/action.yml [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+name: Deploy to Google App Engine
+author: GoogleCloudPlatform
+description: |-
+  App Engine is a fully managed, serverless platform for developing and hosting
```
Comment (craigdbarber) [inline appengine-deploy/action.yml:17]: Consider tailoring the action's description to the functionality provided by the action itself. Perhaps linking to App Engine documentation for a description of the service.

File: appengine-deploy/action.yml [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+      The yaml files for the services or configurations you want to deploy.
+      If not given, defaults to app.yaml in the current directory. If that is
+      not found, attempts to automatically generate necessary configuration
+      files (such as app.yaml) in the current directory.
+    required: false
+
+  image_url:
+    description: |-
+      Deploy with a specific container image. The image url must be from one of
+      the valid gcr hostnames.
```
Comment (craigdbarber) [inline appengine-deploy/action.yml:38]: s/gcr/GCR

File: appengine-deploy/action.yml [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+    description: |-
+      The yaml files for the services or configurations you want to deploy.
+      If not given, defaults to app.yaml in the current directory. If that is
+      not found, attempts to automatically generate necessary configuration
+      files (such as app.yaml) in the current directory.
+    required: false
+
+  image_url:
+    description: |-
+      Deploy with a specific container image. The image url must be from one of
```
Comment (craigdbarber) [inline appengine-deploy/action.yml:37]: s/url/URL

File: setupGcloudSDK/src/index.ts (hunk 5 above)
Comment (craigdbarber) [reply setupGcloudSDK/src/index.ts:151]: Gotcha, disregard :)

File: appengine-deploy/README.md (hunk 6 above)
Comment (averikitsch) [reply appengine-deploy/README.md:22]: done.

File: appengine-deploy/README.md (hunk 7 above)
Comment (averikitsch) [reply appengine-deploy/README.md:28]: done.

File: appengine-deploy/README.md (hunk 8 above)
Comment (averikitsch) [reply appengine-deploy/README.md:81]: done.

File: appengine-deploy/README.md (hunk 9 above)
Comment (averikitsch) [reply appengine-deploy/README.md:92]: done.

File: appengine-deploy/action.yml (hunk 10 above)
Comment (averikitsch) [reply appengine-deploy/action.yml:17]: done.

File: appengine-deploy/action.yml (hunk 11 above)
Comment (averikitsch) [reply appengine-deploy/action.yml:38]: done.

File: appengine-deploy/action.yml (hunk 12 above)
Comment (averikitsch) [reply appengine-deploy/action.yml:37]: done.

**Comment by craigdbarber** [comment]
Ran the TF scripts to create the IT infra, and added the corresponding secrets. Kicked off a new run of the IT workflows, and it looks like they're still failing. PTAL :)

**Comment by averikitsch** [comment]
@craigdbarber  It looks like I forgot to add `secrets.APPENGINE_DEPLOY_PROJECT_ID` to my infra.

**Comment by craigdbarber** [comment]
@averikitsch I added you as an admin to the project. You should be able to add it yourself under Settings -> Secrets.

**Comment by averikitsch** [comment]
@sethvargo the secret managers IT for setup-gcloud are failing when calling acessSecret. Do you have any insight on to why this would be failing when the setup-gcloud is working?

**Pull request merged**